File: app/src/main/java/kravchenko/ru/cleanarchitecture/app/di/RepositoryModule.java
package kravchenko.ru.cleanarchitecture.app.di;

import javax.inject.Named;

import dagger.Module;
import dagger.Provides;
import kravchenko.ru.cleanarchitecture.app.di.annotation.RepositoryProd;
import kravchenko.ru.cleanarchitecture.app.di.annotation.RepositoryTest;
import kravchenko.ru.cleanarchitecture.data.repository.Repository;
import kravchenko.ru.cleanarchitecture.data.repository.TestRepository;
import kravchenko.ru.cleanarchitecture.domain.source.DataSource;

@Module
class RepositoryModule {

    @Provides
    DataSource provideRepository() {
        return new Repository();
    }

}

File: app/src/main/java/kravchenko/ru/cleanarchitecture/app/di/RepositoryModule.kt
package kravchenko.ru.cleanarchitecture.app.di

import dagger.Module
import dagger.Provides
import io.reactivex.Single
import kravchenko.ru.cleanarchitecture.app.di.annotation.MyScope
import kravchenko.ru.cleanarchitecture.data.model.User
import kravchenko.ru.cleanarchitecture.domain.source.DataSource

@Module
internal class RepositoryModule(private val repos: DataSource) {

    @MyScope
    @Provides
    fun provideRepository(): DataSource {
        return repos
    }

    @MyScope
    @Provides
    fun provideUsersUseCase(): () -> Single<List<User>> {
        return repos.getUsers()
    }

}
